[camera] Camera Binding initial contribution

thhart opened the pull request:
This addon offers fetching images of HTTP API based IP cameras.

kaikreuzer commented:
Thanks - please note that you do not need to close and open new PRs if something is wrong; you can always simply do a "force push" to your branch, which will update the existing PR accordingly.

martinvw commented:
Can you do a quick check whether this is the same or similar to https://github.com/openhab/openhab2-addons/pull/2111 Thanks!

martinvw commented:
There are some initial issues already detected by travis:

```
[ERROR] Failed to execute goal org.openhab.tools:static-code-analysis:0.0.4:report (default) on project org.openhab.binding.camera: Code Analysis Tool has found 6 error(s)!
[ERROR] Please fix the errors and rerun the build.
[ERROR] Errors list:
[ERROR] ERROR found by findbugs: org.openhab.binding.camera.handler.CameraHandler.java:60 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by checkstyle: .binding.camera/META-INF/about.html:0 Missing about.html file.
[ERROR] ERROR found by findbugs: org.openhab.binding.camera.handler.CameraHandler.java:87 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.camera.handler.CameraHandler.java:96 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.camera.handler.CameraHandler.java:100 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.camera.handler.CameraHandler.java:104 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
```

thhart commented:
1. https://github.com/openhab/openhab2-addons/pull/2111
The addon of 2111 is a special UPnP camera plugin which is "only" for a special protocol. The addon I publish here has a wider approach which is supporting a lot of HTTP API based cameras like Axis, Bosch and a lot others.
2. I fixed the performance issues and added a short about.hml.

thhart commented:
Sorry, I did not issue a forced push, but I can seem my commit is already mentioned here: https://github.com/openhab/openhab2-addons/pull/2282/commits/2939517ac91cb6667b1f82ad14fc9053fe164534.
Is there anything I should do further?

martinvw commented on addons/binding/org.openhab.binding.camera/.classpath:
```diff
@@ -0,0 +1,11 @@
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.7"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src/main/java"/>
+	<classpathentry exported="true" kind="lib" path="lib/httpcore-4.4.6.jar"/>
```
Would be great to depend on the already provided Jetty instead...

We also have a wrapper that support fetching of images out of the box: 
https://www.eclipse.org/smarthome/documentation/javadoc/org/eclipse/smarthome/io/net/http/HttpUtil.html

thhart replied:
The reason for using the original apache lib here is the special authentication support which I can not see in your HttpUtil. The client really needs all the libs referenced in the manifest.

martinvw replied:
As far I quickly saw this code is just a complex way to add a header:

```
Authorization: Basic base64encode(user+":"+password)
```

So if you would be helped if we add support for additional headers to the `HttpUtil` I would be happy not dragging along these extra dependencies. @kaikreuzer given our pending `HttpUtil` discussion wdyt?

Source: https://en.wikipedia.org/wiki/Basic_access_authentication

martinvw commented on addons/binding/org.openhab.binding.camera/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,11 @@
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+    <name>Camera Binding</name>
+    <description>Binding for generic IP cameras.</description>
```
Would the correct description be: 'Generic binding for IP cameras.'

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/ESH-INF/config/cameraConfig.xml:
```diff
@@ -0,0 +1,33 @@
+        http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
+    <config-description uri="thing-type:camera:camera">
+        <parameter name="poll" unit="s" type="integer">
+            <label>Poll time:</label>
```
I doubt whether the colon should be included here, please validate in other bindings.

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/ESH-INF/config/cameraConfig.xml:
```diff
@@ -0,0 +1,33 @@
+        </parameter>
+        <parameter name="urlVideo" type="text">
+            <label>Video Url:</label>
+            <description>Url for receiving the camera video. ..TODO..</description>
```
There is a TODO, what should be done and can it be completed before merging the PR?

thhart replied:
This is on top of my list but really can not make it into release, I commented out in the meanwhile.

martinvw commented on addons/binding/org.openhab.binding.camera/README.md:
```diff
@@ -0,0 +1,49 @@
+Almost every IP camera has a simple HTTP access API defined by Urls to access at least static images.
+Additionally video streams like H264.x oder MJPEG and others are very wide spread. This binding supports these types of cameras.
+
+-- Thomas Hartwig 2016-07-24
```
What / why is this?

thhart replied:
OK, changed the text.

martinvw commented on addons/binding/org.openhab.binding.camera/README.md:
```diff
@@ -0,0 +1,49 @@
+
+## Thing Configuration
+
+* self explaining when devices are created in OpenHab
```
And add a full example

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/README.md:
```diff
@@ -0,0 +1,49 @@
+
+This is the base image channel which is updated periodically in the configured interval. This is currently the only supported channel.
+
+## Near Future
```
I think you can remove this from the readme, the goal of the readme is different.

thhart replied:
OK, moved it to a TODO file.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+    private final AtomicBoolean initialized = new AtomicBoolean(false);
+    private final ExecutorService serviceCached = Executors.newCachedThreadPool();
+    private String urlSnapshot;
+    private String urlUsername = "";
```
Should it really be initialized with an empty string?

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+ * @author Thomas Hartwig - Initial contribution
+ */
+public class CameraHandler extends BaseThingHandler {
+    private Logger logger = LoggerFactory.getLogger(CameraHandler.class);
```
I like loggers `final`

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(CHANNEL_IMAGE)) {
+            if (command.toString().equals("REFRESH")) {
```
Please check other bindings how to check for the refresh command

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+        if (logger.isDebugEnabled()) {
+            logger.debug("Initialize thing: {}::{}", getThing().getLabel(), getThing().getUID());
+        }
+        Object paramUrl = getConfig().get("urlSnapshot");
```
This is already a `String` and can just be casted.

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+        }
+        Object paramUrl = getConfig().get("urlSnapshot");
+        urlSnapshot = String.valueOf(paramUrl);
+        Object paramUsername = getConfig().get("urlUsername");
```
See above

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+        if (paramUsername != null) {
+            urlUsername = String.valueOf(paramUsername);
+        }
+        Object paramPassword = getConfig().get("urlPassword");
```
See above

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+        }
+        try {
+            Object param = getConfig().get("poll");
+            CONFIGURATION_POLLTIME_MS = (int) (Double.parseDouble(String.valueOf(param)) * 1000);
```
I suppose the type of this is a BigDecimal, you can just cast it and ask the `intValue` of it. Maybe a null check could be wise :-)

Also do not overwrite constants, store the polltime here in this class and just the constant as the default value if needed.

thhart replied:
Changed according to polltime fix.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+                    while (weakReference.get() != null) {
+                        try {
+                            refreshData();
+                        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

thhart replied:
OK, removed the catch.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+                        } catch (Exception e) {
+                            logger.error("error in refresh", e);
+                        }
+                        Thread.sleep(Math.max(10, CONFIGURATION_POLLTIME_MS));
```
This is not the way, please use a scheduler, which is available from the parent class, you can then schedule it to run every polltime. See examples in other bindings. Questions, let me know!

thhart replied:
I have a problem using a scheduler here. As I have learned a binding might be reconfigured and initialized during a reconfiguration process without being reinstantiated*. The reconfiguration can affect my polling. This is not possible when using the parent scheduler.scheduleAtFixedRate. Of course I can submit my callable instead to my service to the parent scheduler with my poll time sleep which was reread from configuration. Is this better?

*) I am not right sure if this really is true. If it is the case many other bindings might be wrong and use the scheduler wrong then. This is why I used the initializer atomic here.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+                for (Channel cx : getThing().getChannels()) {
+                    if (cx.getAcceptedItemType().equals("Image")) {
+                        if (logger.isTraceEnabled()) {
+                            logger.trace("Will update: {}::{}::{}", getThing().getUID().getId(),
```
The check is not needed since there are no expensive operations involved.

thhart replied:
OK

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+                                    cx.getChannelTypeUID().getId(), getThing().getLabel());
+                        }
+                        if (urlSnapshot != null) {
+                            try {
```
Rather split this complete method of and consider the using of the `HttpUtil` as mentioned earlier.

thhart replied:
See comment below.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/internal/CameraHandlerFactory.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright (c) 2014 openHAB UG (haftungsbeschraenkt) and others.
```
It seems that this file has a missing, outdated or otherwise non standard copyright header, can you fix all of them by running 

`mvn license:format` 

Please commit only the ones for your binding.

thhart replied:
OK

thhart replied:
@martinvw When I run 'mvv license:format' in the openhab2-addons base dir it wants to download the whole eclipse history at extreme slow speed, what is it doing there and can I shorten this?

thhart replied:
OK, found a way. Headers now adjusted.

thhart commented on addons/binding/org.openhab.binding.camera/ESH-INF/config/cameraConfig.xml:
```diff
@@ -0,0 +1,33 @@
+        <parameter name="urlPassword" type="text">
+            <label>Password:</label>
+            <description>Password for accessing camera (empty for standard)</description>
+            <default></default>
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/ESH-INF/i18n/camera_xx_XX.properties:
```diff
@@ -0,0 +1,11 @@
+# binding
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/ESH-INF/thing/camera.xml:
```diff
@@ -0,0 +1,22 @@
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+    <!-- Sample Thing Type -->
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/ESH-INF/thing/camera.xml:
```diff
@@ -0,0 +1,22 @@
+        <config-description-ref uri="thing-type:camera:camera" />
+    </thing-type>
+
+    <!-- Sample Channel Type -->
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+Bundle-Version: 2.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .,
+ lib/commons-codec-1.10.jar,
```
Yes, see below regarding the HTTPUtil.

thhart commented on addons/binding/org.openhab.binding.camera/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+ lib/httpcore-4.4.6.jar,
+ lib/httpmime-4.5.2.jar
+Import-Package: 
+ org.apache.commons.io;version="2.2.0",
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/README.md:
```diff
@@ -0,0 +1,49 @@
+All cameras supporting simple HTTP based API fetching for JPG/PNG images.
+Please contact the author Thomas Hartwig <thomas.hartwig@gmail.com> for camera Url examples if not listed here.
+
+Vendor | Single snapshot URL | Video capture (not yet supported) |
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/about.html:
```diff
@@ -0,0 +1,44 @@
+  
+    </p>    
+</body>
+</html>
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/CameraBindingConstants.java:
```diff
@@ -0,0 +1,29 @@
+    public static final String BINDING_ID = "camera";
+
+    // List of all Thing Type UIDs
+    public final static ThingTypeUID THING_TYPE_CAMERA = new ThingTypeUID(BINDING_ID, "camera");
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/CameraBindingConstants.java:
```diff
@@ -0,0 +1,29 @@
+
+    // List of all Channel ids
+    public final static String CHANNEL_IMAGE = "image";
+    public static int CONFIGURATION_POLLTIME_MS = 10;
```
OK, fixed it and removed the CONFIGURATION_POLLTIME_MS. Polling is 5 seconds by default now.

thhart commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(CHANNEL_IMAGE)) {
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+        if (logger.isDebugEnabled()) {
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,195 @@
+            });
+        }
+        updateStatus(ThingStatus.OFFLINE);
+        // Note: When initialization can NOT be done set the status with more details for further
```
OK

thhart commented on addons/binding/pom.xml:
```diff
@@ -24,6 +24,7 @@
     <module>org.openhab.binding.avmfritz</module>
     <module>org.openhab.binding.bigassfan</module>
     <module>org.openhab.binding.boschindego</module>
+    <module>org.openhab.binding.camera</module>
```
OK

thhart commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/CameraBindingConstants.java:
```diff
@@ -0,0 +1,29 @@
+    public final static ThingTypeUID THING_TYPE_CAMERA = new ThingTypeUID(BINDING_ID, "camera");
+
+    // List of all Channel ids
+    public final static String CHANNEL_IMAGE = "image";
```
OK

martinvw commented:
You seriously messed up your git repo, can you fix it, please.

Please that this will not happen so easily of you follow some simple rules, do not merge master in your repo but instead rebase. And use feature branch in your repo instead of developing on the master.

Not that you can force push to overwrite the branch that this PR is linked to, it would help me in keeping tracking of review comments I already submitted.

thhart commented:
@martinvw Sorry for the noise, I forced now. Should be better now.
I do not understand fully, should I use a specific "feature" branch already present or should I create a new one?

martinvw commented:
Looks way better now, thanks. i will try to check it out later today!

martinvw commented on addons/binding/org.openhab.binding.camera/README.md:
```diff
@@ -0,0 +1,38 @@
+
+* Poll time: the update interval of the image to fetch from the camera
+* Snapshot Url: the url to fetch the image from see examples above
+* Username: username when the camera is protected, this uses the standard HTTP authorization protocol,
```
Should a part of this be present in UI, so in the XML, wdyt?

martinvw commented on addons/binding/org.openhab.binding.camera/README.md:
```diff
@@ -0,0 +1,38 @@
+* Poll time: the update interval of the image to fetch from the camera
+* Snapshot Url: the url to fetch the image from see examples above
+* Username: username when the camera is protected, this uses the standard HTTP authorization protocol,
+however username and password might be specified via the Snapshot Url as a parameter, please check camera model documentation for correct parameter syntax
```
Minor: can you write URL as URL 😄

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (CHANNEL_IMAGE.equals(channelUID.getId())) {
+            if (command instanceof RefreshType) {
+                refreshData();
```
Given you have to fetch it from some camera url I would expect to run this in a scheduler.

martinvw replied:
Maybe just:

```java
scheduler.schedule(this::refreshData, 0, TimeUnit.SECONDS);

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+        urlUsername = (String) getConfig().get("urlUsername");
+        urlPassword = (String) getConfig().get("urlPassword");
+        try {
+            Object param = getConfig().get("poll");
```
Imho, this should be a BigDecimal already, also there is no reason to convert it ms

martinvw replied:
You should however have a null check and fallback to a default in that case.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+            logger.warn("could not read poll time from configuration", e1);
+        }
+        logger.debug("Schedule update at fixed rate {} ms.", polltime_ms);
+        if (initialized.compareAndSet(false, true)) {
```
Please get rid of this complex logic, use something like all other bindings do:

```
scheduler.scheduleWithFixedDelay(this::refreshData, 0, polltime, TimeUnit.SECONDS);
```

martinvw replied:
And don't forget to remove the scheduled job from the dispose method, see other bindings for examples.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+        try {
+            Object param = getConfig().get("poll");
+            polltime_ms = (long) (Double.parseDouble(String.valueOf(param)) * 1000);
+        } catch (Exception e1) {
```
No reason to catch exceptions here, it's just a `BigDecimal` or `null`, otherwise it is a framework bug and you should report it.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    private void refreshData() {
+        if (refreshInProgress.compareAndSet(false, true)) {
```
If you want to guard against this, good idea, please take a look at `ExpiringCache` its part of Eclipse SmartHome since a few months and it should resolve this way better.

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+                    if ("Image".equals(cx.getAcceptedItemType())) {
+                        logger.trace("Will update: {}::{}::{}", getThing().getUID().getId(),
+                                cx.getChannelTypeUID().getId(), getThing().getLabel());
+                        if (urlSnapshot != null) {
```
If this is not set you can immediately leave the whole method, maybe start the method with:

```java
if (urlSnapshot == null) {
    // update status
    return;
}
```

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+                                        "unknown error: " + e.toString());
+                            }
+                        } else {
+                            updateStatus(ThingStatus.UNINITIALIZED, ThingStatusDetail.CONFIGURATION_PENDING);
```
Is this the correct status, or maybe:

```ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR```

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+                                cx.getChannelTypeUID().getId(), getThing().getLabel());
+                        if (urlSnapshot != null) {
+                            try {
+                                final URL url = new URL(urlSnapshot);
```
Leave out the `final`

martinvw commented on addons/binding/org.openhab.binding.camera/src/main/java/org/openhab/binding/camera/handler/CameraHandler.java:
```diff
@@ -0,0 +1,177 @@
+        }
+    }
+
+    private static ByteArrayOutputStream readImage(URL url, String username, String password) throws Exception {
```
as mentioned on the dependencies.

kaikreuzer commented on addons/binding/org.openhab.binding.camera/README.md:
```diff
@@ -0,0 +1,38 @@
+## Supported Cameras
+
+All cameras supporting simple HTTP based API fetching for JPG/PNG images.
+Please contact the author Thomas Hartwig <thomas.hartwig@gmail.com> for camera Url examples if not listed here.
```
I would completely refrain from putting such information in the README as this is the source for the official docs website. And the mechanism is the very same for all bindings: For any updates/reports/requests the issue tracker should be used - so no need to repeat this here.

martinvw replied:
Agreed

davidgraeff commented:
@thhart Can you confirm that you are not working on this one anymore?

thhart commented:
> @thhart Can you confirm that you are not working on this one anymore?

Hi David, I confirm I am not working on it any more.
